### PR TITLE
dev: directory node setup for regtest

### PR DESCRIPTION
**Pull request by theborakompanioni**
Sets up a local directory node in regtest environment.

### Rationale
The test system should mirror the live system in most regards as closely as possible.

#### Benefits
:+1: Increased Reliability; IRC communication was cumbersome
:+1: Lower latency; Increased execution time
:+1: Closer to the mainnet system; communication is done over tor

#### Drawbacks
:-1: Longer build time (own image with `--with-local-tor` must be built)
:-1: Needs a prepare script before setup can boot up
:-1: Communication is done over tor; might not be good/fast for everyone
:-1: Internet connection needed

### Approach
Problem: We do not know the hostname of the directory node before the setup is started.

Solution: Before the setup can be started it is prepared via script `prepare-setup.sh`.
This in turn calls `generate-onion-address.sh` to generate onion key pair + hostname which will be written to an `.env.generated` file. The env file is used in `docker-compose.yml` to inject the hostname to all JM instances.

The start script `start-dn.py` in the docker image is taken from repository [`JoinMarket-Org/custom-scripts` ](https://github.com/JoinMarket-Org/custom-scripts/blob/c8c3bbb3aeaa86e70b08e7d655e5ed676ff31b8f/start-dn.py).

#### Caveats
- a `.env.example` file exists on purpose to be able to execute commands before the `.env.generated` has been created. This is necessary if `regtest:up` has never been invoked, but other actions should still work (e.g. `regtest:rebuild`, `regtest:clear`, etc.)

- "Cheating" a little, when creating onion keys: Generating the hostname is done with a fork of [`Kexkey/vanitytorgen`](https://github.com/Kexkey/vanitytorgen). This was the easiest and quickest method available. Although purposed to generate vanity addresses, by choosing a random starting char it generates unique hostnames without prefix at every start-up (refrained from using "jam" as prefix). This can be improved – feedback welcome.

### Testing
- Rebuild the setup with `npm run regtest:rebuild`
- Start with `npm run regtest:up`
- Follow outputs of log file with `npm run regtest:logs:directory_node`
- Do whatever in the primary or secondary instance and observe if your actions are successful

###### Current additional Changes (that will likely be reverted and applied in a different PR, as they do not fit the primary objective)
Moved to https://github.com/joinmarket-webui/joinmarket-webui/pull/268
- ~~Set the same fees used as defaults in the jam "production" image~~
- ~~Because the communication is more reliable with directory nodes, let's also:~~
  - ~~Revert the `taker_utxo_retries` from 10 to 3 (default value)~~
  - ~~Revert the `taker_utxo_age` from 1 to 5 (default value)~~

It has been tested with https://github.com/joinmarket-webui/joinmarket-webui/pull/268 and it seems to work fine.
"Schedule said to process "6 transactions over 23 seconds" and then took ~7 minutes with block time at 10 seconds."
Tests included: Running scheduler (with "insecure test options"), a single taker operation, and a successful maker operation.

### CODE REVIEW

**Review by dnlggr — COMMENTED** [review]
Seems to work fine for me! Haven't looked at the code yet, though.

But I tested:

- Scheduler
- Earn
- Send with 1 collaborator

How will the messaging server be chosen (onion vs. IRC) -- randomly?

**Comment by theborakompanioni** [comment]
> 


Thanks for your comments so far and for testing :+1:
> Seems to work fine for me! white_check_mark Haven't looked at the code yet, though.

:muscle: 

> But I tested:
> 
>     * Scheduler
> 
>     * Earn
> 
>     * Send with 1 collaborator

:+1:
Did you notice an increase in speed and reliability?

> How will the messaging server be chosen (onion vs. IRC) -- randomly?

I have not looked into the code and could not find explanations regarding this in https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/master/docs/onion-message-channels.md – it seems that onion messaging takes precedence, but this is in uninformed guess.

File: docker/regtest/dockerfile-deps/joinmarket/directory_node/autostart [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# Remove comments in from of the service you want to automatically restart
```
[the remark this answers is not in the document]
Comment (theborakompanioni) [reply docker/regtest/dockerfile-deps/joinmarket/directory_node/autostart]: Done.

File: docker/regtest/dockerfile-deps/joinmarket/directory_node/README.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+e.g. in a docker-compose setup:
+```yml
+  joinmarket_directory_node:
+    [...]
+    environment:
+      jm_hidden_service_dir: \/root\/.joinmarket\/hidden_service_dir
+      jm_directory_nodes: jamobvon3jdgjnvuxur7vpdr6hzwrxhu5li3pnannfrrrupg5sb6ouyd.onion:5222
+    volumes:
+      - "./my-hidden_service_dir:/root/.joinmarket/hidden_service_dir:z"
+
```
[the remark this answers is not in the document]
Comment (theborakompanioni) [reply docker/regtest/dockerfile-deps/joinmarket/directory_node/README.md]: Done.

**Comment by dnlggr** [comment]
Just tried it again. Worked reliably. So while I can't know if it is more reliable or faster it feels good. 👌

**Review by dnlggr — APPROVED** [review]

**Pull request merged**